### DISCUSSION TITLE
JavaScript templating libraries.

### DISCUSSION THREAD

**Discussion by dj2** [cds-snc/docs#46] [Tips and Techniques]
Was trying a few of these and wanted to write down some notes. For the accessibility tool we ended up with Nunjucks which is also what the node starter app used.

# Pug
We started with pug. For many of us the syntax is hard to read along with having some strange conflicts with tailwind. When you have classes in tailwind like `mx:2` or any style with strange characters you end up with pug like `.foo.class(mx:2)` which is, strange to read. I was also unable to get interpolation working inside attributes to tags for things like getting the locale for i18n.

# Handlebars
Handlebars worked and is primary HTML so was easy to read. The issue came when we started doing i18n. You attach the `__` method to the response. But, if you have a loop inside the template, that gets a new context so the i18n methods aren't attached. You end up having to do `{{ ../__ "foo" }}`, so walk the context path which feels weird.

# Nunjucks
So far, nunjucks is able to handle i18n in a fairly easy fashion, is HTML so easier to read.

**Comment by CalvinRodo** [comment]
Personally I've enjoyed using Nunjucks in the past and found it worked quite well and was an easy lift for someone that was used to the templating languages in the .Net and Java world.

Although to be honest the other two aren't difficult either.

**Reply by dj2** [reply]
The biggest issue I had with the other two was getting them to work with multiple languages. Nunjucks was, by far, the easiest  here as it just worked.
